### PR TITLE
Service Account auth from saved credentials; Fix and run tests on Travis.

### PR DESCRIPTION
**Pull request by MaxRis**

### CODE REVIEW

File: pydrive2/auth.py [hunk 1]
```diff
@@ -273,12 +273,13 @@ def ServiceAuth(self):
     user_email = self.client_config.get('client_user_email')
     service_email = self.client_config['client_service_email']
     file_path = self.client_config['client_pkcs12_file_path']
-    self.credentials = ServiceAccountCredentials.from_p12_keyfile(
-        service_account_email=service_email,
-        filename=file_path,
-        scopes=scopes)
-    if user_email:
-        self.credentials = self.credentials.create_delegated(sub=user_email)
+    if file_path:
```
Comment (shcheklein) [inline pydrive2/auth.py]: could you please elaborate on this? why do we need this check?

File: pydrive2/auth.py [hunk 2]
```diff
@@ -523,7 +524,7 @@ def Authorize(self):
     """
     if self.http is None:
       self.http = httplib2.Http(timeout=self.http_timeout)
-    if self.access_token_expired:
+    if self.access_token_expired and self.auth_method != 'service':
```
Comment (shcheklein) [inline pydrive2/auth.py]: same here, not sure I understand why do we need this second check

File: pydrive2/auth.py (hunk 1 above)
Comment (MaxRis) [reply pydrive2/auth.py]: if `file_path` is not specified it shouldn't try to load p12 key from unspecified path and error out. The idea is that encrypted key and secret to access service account already might be stored in presaved credentials file generated from previous app run.

Comment (shcheklein) [reply pydrive2/auth.py]: so, it's still required for the first run?

Comment (shcheklein) [reply pydrive2/auth.py]: in this case this method does not do anything at all?

Comment (MaxRis) [reply pydrive2/auth.py]: These changes were made to allow to run gdrive tests on Travis only with specified GDRIVE_USER_CREDENTIALS_DATA secured env var and service account email in settings.
 
> in this case this method does not do anything at all?

You are right. Except validation of available settings by call of `self.LoadServiceConfigSettings()` it seems doesn't do anything else. 
I guess `if file_path:` can be moved right after call of `self.LoadServiceConfigSettings()`

Comment (MaxRis) [reply pydrive2/auth.py]: > so, it's still required for the first run?

Yes, it is required for the first run (except DVC specific scenario if GDRIVE_USER_CREDENTIALS_DATA env var is set from the first run ). Might makes sense to check there if credentials file can be found at all to not proceed further if it doesn't exist yet.

File: pydrive2/auth.py (hunk 2 above)
Comment (MaxRis) [reply pydrive2/auth.py]: "For the Service account we don't want to Refresh the access token, since refresh token is not available for Service accounts. We just want to Auth as usual and request new access token."

File: pydrive2/auth.py (hunk 1 above)
Comment (shcheklein) [reply pydrive2/auth.py]: if it does not do anything, why is it being called at all?

File: pydrive2/auth.py (hunk 2 above)
Comment (shcheklein) [reply pydrive2/auth.py]: do we need a p12 file around for that?

File: pydrive2/auth.py (hunk 1 above)
Comment (MaxRis) [reply pydrive2/auth.py]: @shcheklein I've added `return` in case if credentials were already loaded from the file. Should work now in similar way as auth for usual Google account.

File: pydrive2/auth.py (hunk 2 above)
Comment (MaxRis) [reply pydrive2/auth.py]: at this stage the content of p12 file already should be loaded into `self.credentials`. If it was loaded from `gdrive-user-credentials.json` then p12 file is not expected to be used and the path is not verified.

File: pydrive2/auth.py [hunk 3]
```diff
@@ -521,12 +521,14 @@ def Authorize(self):
 
     :raises: AuthenticationError
     """
+    if self.access_token_expired and self.auth_method != 'service':
```
Comment (shcheklein) [inline pydrive2/auth.py]: so, we don't need this additional check anymore? service credentials do have `access token` and it should not be expired at this point unless I'm missing something. There should no be difference between workflows in this case.

File: pydrive2/auth.py [hunk 4]
```diff
@@ -521,12 +521,14 @@ def Authorize(self):
...
+      raise AuthenticationError('No valid credentials provided to authorize')
+
     if self.http is None:
       self.http = httplib2.Http(timeout=self.http_timeout)
-    if self.access_token_expired:
-      raise AuthenticationError('No valid credentials provided to authorize')
     self.http = self.credentials.authorize(self.http)
-    self.service = build('drive', 'v2', http=self.http, cache_discovery=False)
+    if self.service is None:
```
Comment (shcheklein) [inline pydrive2/auth.py]: same here - I feel that this is unrelated change

File: pydrive2/auth.py [hunk 5]
```diff
@@ -95,11 +95,8 @@ def _decorated(self, *args, **kwargs):
       dirty = True
     else:
       if self.access_token_expired:
-        if self.credentials.refresh_token is not None:
```
Comment (shcheklein) [inline pydrive2/auth.py]: so, the code a few lines above - why do we call `Authorize`? and how does Authorize responsible for building the credentials at all?

Comment (shcheklein) [reply pydrive2/auth.py]: if credentials are none for some reason we need to call `decoratee` like in the CheckAuth below?

it feels like it should be very similar to the CheckAuth (with some modification related to refresh token, and `service` auth type)

File: pydrive2/auth.py [hunk 6]
```diff
@@ -95,11 +95,8 @@ def _decorated(self, *args, **kwargs):
...
-          self.Refresh()
-        else:
-          decoratee(self, *args, **kwargs)
-          self.Authorize()
+        decoratee(self, *args, **kwargs)
```
Comment (shcheklein) [inline pydrive2/auth.py]: you probably don't need to call decoratee here at all, and then it means you don't to include that artificial if in the `ServiceAuth` code

File: pydrive2/auth.py (hunk 3 above)
Comment (MaxRis) [reply pydrive2/auth.py]: right, makes sense to remove this check

File: pydrive2/auth.py (hunk 5 above)
Comment (MaxRis) [reply pydrive2/auth.py]: good catch, it makes sense, thanks!

**Comment by shcheklein** [comment]
looks good! did you manage to run tests that are included in the library? can we fix/add a test for this?

**Comment by MaxRis** [comment]
@shcheklein some tests are failing, will look into it

File: pydrive2/test/test_util.py [hunk 7]
```diff
@@ -1,9 +1,53 @@
...
+        ) as credentials_file:
+            credentials_file.write(
+                os.getenv(GDRIVE_USER_CREDENTIALS_DATA)
+            )
+
+class GDriveRetriableError(Exception):
+    pass
+
+@decorator
+def _wrap_pydrive_retriable(call):
```
Comment (shcheklein) [inline pydrive2/test/test_util.py]: PTAL at the latest cleanup by @Suor in the main repo - he has wrapped it nicely with some nice new features from `fancy`

Comment (MaxRis) [reply pydrive2/test/test_util.py]: So far, I've left implementation as is (it was changed and simplified since this your comment). The difference with DVC is in usage: in-lined lambda vs decorator.

File: .travis.yml [hunk 8]
```diff
@@ -11,7 +11,7 @@ env:
     - PIP_CACHE_DIR="$HOME/.cache/pip"  # unify pip cache location for all platforms
 stages:
 #  - check
```
Comment (shcheklein) [inline .travis.yml]: minor: let's remove unused stages

File: .travis.yml [hunk 9]
```diff
@@ -11,7 +11,7 @@ env:
...
-#  - test
+  - test
   - build
```
Comment (shcheklein) [inline .travis.yml]: I suppose we don't support publishing yet - it should be failing, no?

File: pydrive2/test/test_util.py [hunk 10]
```diff
@@ -1,9 +1,60 @@
...
+from funcy.py3 import cat
+from pydrive2.files import ApiRequestError
 
 newline_pattern = re.compile(r'[\r\n]')
 
 
+class PyDriveTestBase(unittest.TestCase):
+    GDRIVE_USER_CREDENTIALS_DATA = "GDRIVE_USER_CREDENTIALS_DATA"
+    DEFAULT_USER_CREDENTIALS_FILE = "credentials/1.dat"
+    if os.getenv(GDRIVE_USER_CREDENTIALS_DATA):
```
Comment (shcheklein) [inline pydrive2/test/test_util.py]: should it be part of setUp or something? it is strange to see so much code on the global class level

File: pydrive2/test/test_util.py [hunk 11]
```diff
@@ -1,9 +1,60 @@
...
+            credentials_file.write(
+                os.getenv(GDRIVE_USER_CREDENTIALS_DATA)
+            )
+
+class PyDriveRetriableError(Exception):
+    pass
+
+
+# 15 tries, start at 0.5s, multiply by golden ratio, cap at 20s
+@retry(15, PyDriveRetriableError, timeout=lambda a: min(0.5 * 1.618 ** a, 20))
```
Comment (shcheklein) [inline pydrive2/test/test_util.py]: @Suor can we use the simplified version from the latest DVC core PR here?

File: pydrive2/test/test_file.py [hunk 12]
```diff
@@ -493,9 +521,10 @@ def test_Gfile_Conversion_Add_Remove_BOM(self):
     finally:
       self.cleanup_gfile_conversion_test(file1, file_name, downloaded_file_name)
 
+
   def test_InsertPrefix(self):
     # Create BytesIO.
-    file_obj = BytesIO('abc')
+    file_obj = BytesIO('abc'.encode('utf8'))
```
Comment (shcheklein) [inline pydrive2/test/test_file.py]: is it py2 vs py3 difference?

File: pydrive2/test/test_oauth.py [hunk 13]
```diff
@@ -3,10 +3,15 @@
...
+from pydrive2.test import test_util
 
-class GoogleAuthTest(unittest.TestCase):
+class GoogleAuthTest(test_util.PyDriveTestBase):
   """Tests basic OAuth2 operations of auth.GoogleAuth."""
 
+  def setUp(self):
+    if os.getenv(test_util.PyDriveTestBase.GDRIVE_USER_CREDENTIALS_DATA):
+      self.skipTest('OAuth test is skipped. Unset {} env var.'.format(test_util.PyDriveTestBase.GDRIVE_USER_CREDENTIALS_DATA))
+
```
Comment (shcheklein) [inline pydrive2/test/test_oauth.py]: I probably don't understand the whole logic behind test1 - test6 files, but should we use test6 for ServiceAuth ?

Comment (shcheklein) [reply pydrive2/test/test_oauth.py]: why do we bother here at all if we just skip these tests anyway? is there a way to run them?

File: .travis.yml (hunk 8 above)
Comment (MaxRis) [reply .travis.yml]: it might makes sense to setup stage with code formatting check as well

File: .travis.yml (hunk 9 above)
Comment (MaxRis) [reply .travis.yml]: apparently, @efiop have setup publishing already https://github.com/iterative/PyDrive2/commit/d44c23c7ed9df424e9199ae74bcc48a51eccaf70 to run the corresponding script https://github.com/iterative/PyDrive2/blob/master/scripts/build_package.sh#L12

File: pydrive2/test/test_util.py (hunk 10 above)
Comment (MaxRis) [reply pydrive2/test/test_util.py]: right, have updated all automated TestCase classes to do setup in
```
@classmethod
  def setup_class(cls):
```

File: pydrive2/test/test_file.py (hunk 12 above)
Comment (MaxRis) [reply pydrive2/test/test_file.py]: Not sure from where it comes exactly, but `BytesIO` doesn't accept plain string anymore and complains with error `TypeError: a bytes-like object is required, not 'str'`

**Comment by shcheklein** [comment]
@MaxRis looks like a bunch of png1, png2 files were committed, do we need them?

File: .travis.yml [hunk 14]
```diff
@@ -10,45 +10,28 @@ env:
...
   include:
-  # patch formatting checks
-#  - name: "Check patch formatting"
-#    stage: check
-#    os: linux
-#    language: python
-#    python: 3.7
-#    before_install:
-#    install:
-#    script: ./scripts/ci/check_patch.sh
```
Comment (efiop) [inline .travis.yml]: How about we enable patch style checks too? Black should help keeping the style consistent and flake8 will point out some issues. Maybe even pylint might be easy to enable at this point. Maybe not in this PR, but just wanted to mention this.

File: pydrive2/test/test_apiattr.py [hunk 15]
```diff
@@ -2,20 +2,27 @@
...
-    second_file = 'b.png'
+
+    first_file = 'pydrive2/test/files/a.png'
+    second_file = 'pydrive2/test/files/b.png'
+
+    @classmethod
+    def setup_class(cls):
+        setup_credentials()
+
+        cls.ga = GoogleAuth('pydrive2/test/settings/test1.yaml')
```
Comment (shcheklein) [inline pydrive2/test/test_apiattr.py]: why do we need to share this on the class level?

File: pydrive2/test/test_file.py (hunk 12 above)
Comment (shcheklein) [reply pydrive2/test/test_file.py]: kk 👍

File: setup.py [hunk 16]
```diff
@@ -20,6 +20,6 @@
         "pyOpenSSL >= 19.1.0"
     ],
     extras_require={
-        "tests": ["timeout-decorator"],
+        "tests": ["pytest", "timeout-decorator", "funcy>=1.14"],
```
Comment (efiop) [inline setup.py]: Let's add a stronger pytest requirement, will make our lives easier in the future :)

**Comment by MaxRis** [comment]
> Guys, do we really need binary png files here? They will leave in the history forever.

these 2 files - "a.png" and "b.png" are expected by tests to be in place, but it seems we can replace them with any dynamically generate pair of different files.

File: pydrive2/test/test_apiattr.py (hunk 15 above)
Comment (MaxRis) [reply pydrive2/test/test_apiattr.py]: I've moved it into the `setUp` method, since `drive` instance requires it and is supposed to be used by all test methods.

In other TestCase classes `drive` instance usually is not shared across all test methods and separate instance is created per each test method. Not sure if it makes any difference with sharing of single `drive` instance.

File: pydrive2/test/test_oauth.py (hunk 13 above)
Comment (MaxRis) [reply pydrive2/test/test_oauth.py]: Have removed this check completely.
New test for ServiceAuth() and test6.yaml is added.
`@pytest.mark.automated` custom mark is used to run tests on Travis.

File: .travis.yml (hunk 14 above)
Comment (efiop) [reply .travis.yml]: @MaxRis @shcheklein what do you think, guys?

Comment (shcheklein) [reply .travis.yml]: that would super helpful! I would just want to keep the scope of this one reasonable (we already went from service accounts support into implementing Travis) let's make a separate ticket for this?

Comment (efiop) [reply .travis.yml]: Yeah, sure.

**Comment by MaxRis** [comment]
@efiop @shcheklein formatting checks are added in separate commit, should be good to merge all together

**Comment by shcheklein** [comment]
@MaxRis could we please revert the formatting? It's impossible now to review anything. We can apply it later. That why I wanted to make it as a separate step. We have way too many things at once in this PR now.

**Comment by efiop** [comment]
@MaxRis @shcheklein Sorry for the confusion, guys :slightly_frowning_face:

**Comment by MaxRis** [comment]
> @MaxRis could we please revert the formatting? It's impossible now to review anything. We can apply it later. That why I wanted to make it as a separate step. We have way too many things at once in this PR now.

@shcheklein I can revert last commit, no problem, but idea was that all source code changes not related to the formatting cheks are only in the first commit https://github.com/iterative/PyDrive2/pull/11/commits/bcb5e2bb5623ee80a2b0cf5c3d8325fbab86046e which you can review and add comments by above link.

**Comment by shcheklein** [comment]
@MaxRis ah, okay, I see. Will it be easy for you to support this structure going forward? Up to you.

**Comment by MaxRis** [comment]
@shcheklein I supposed only minor further changes to the codebase in scope of this PR and make 3rd commit with them :) But if there will be a lot things to change, then better to revert formatting check commit :)

**Comment by shcheklein** [comment]
@MaxRis sure, let me review the commit then and we can decide after that :)

File: pydrive2/test/settings/test1.yaml [hunk 17]
```diff
@@ -1,9 +1,12 @@
-client_config_backend: 'file'
-client_config_file: 'configs/client_secrets.json'
+client_config_backend: file
```
Comment (shcheklein) [inline pydrive2/test/settings/test1.yaml]: I still think we should not be changing test1.yaml ... it was made to test a different workflow

File: pydrive2/test/test_apiattr.py [hunk 18]
```diff
@@ -1,34 +1,39 @@
...
+    @classmethod
+    def setup_class(cls):
+        setup_credentials()
 
     def test_UpdateMetadataNotInfinitelyNesting(self):
         # Verify 'metadata' field present.
         self.assertTrue(self.file1.metadata is not None)
-        self.file1.UpdateMetadata()
-        self.file1.UpdateMetadata()
+        pydrive_retry(lambda: self.file1.UpdateMetadata())
```
Comment (shcheklein) [inline pydrive2/test/test_apiattr.py]: can we just pass `self.file1.UpdateMetadata` instead of a lambda?

File: pydrive2/test/test_apiattr.py [hunk 19]
```diff
@@ -1,34 +1,39 @@
 import unittest
+import pytest
 
 from pydrive2.auth import GoogleAuth
 from pydrive2.drive import GoogleDrive
+from pydrive2.test.test_util import pydrive_retry, setup_credentials
 
+
+@pytest.mark.automated
 class ApiAttributeTest(unittest.TestCase):
```
Comment (shcheklein) [inline pydrive2/test/test_apiattr.py]: why don't we use `GoogleDriveTest` here?

Comment (shcheklein) [reply pydrive2/test/test_apiattr.py]: or some other generic test that deals with the regular auth that is common for most of the test as far as I understand?

File: pydrive2/test/test_oauth.py [hunk 20]
```diff
@@ -11,7 +13,7 @@ def test_01_LocalWebserverAuthWithClientConfigFromFile(self):
     # Delete old credentials file
     self.DeleteOldCredentialsFile('credentials/1.dat')
     # Test if authentication works with config read from file
-    ga = GoogleAuth('settings/test1.yaml')
+    ga = GoogleAuth('pydrive2/test/settings/test1.yaml')
```
Comment (shcheklein) [inline pydrive2/test/test_oauth.py]: does it still test what it should be testing? (since we changed to p12)

File: pytest.ini [hunk 21]
```diff
@@ -0,0 +1,3 @@
+[pytest]
+markers =
+    automated: mark a test as an automated and runnable on CI.
```
Comment (shcheklein) [inline pytest.ini]: newline

Comment (shcheklein) [reply pytest.ini]: is it better to mark some tests `local` ? it looks like majority (and it should be this way) of test we run automatically

File: pydrive2/test/test_util.py [hunk 22]
```diff
@@ -10,4 +58,19 @@ def CreateRandomFileName():
...
-    return newline_pattern.sub("", string)
\ No newline at end of file
+    return newline_pattern.sub("", string)
+
+
+def create_file(path, content):
+    try:
+        with open(path, "w") as f:
+            f.write(content)
+    except OSError:
```
Comment (Suor) [inline pydrive2/test/test_util.py]: Is it a good idea to just ignore this? This will make the following errors cryptic.

File: setup.py [hunk 23]
```diff
@@ -1,25 +1,36 @@
+import sys
 from setuptools import setup
 
+# Extra dependecies to run tests
+tests_requirements = [
+    "pytest>=4.6.0",
+    "timeout-decorator",
+    "funcy>=1.14",
```
Comment (Suor) [inline setup.py]:
```suggestion
    "funcy>=1.14,<2",
```
If I'll break backwards compatibility. Same for `pytest` I guess.

File: pydrive2/test/test_util.py (hunk 11 above)
Comment (Suor) [reply pydrive2/test/test_util.py]: Yes, we can. However, this is not a decorator, this is direct invoke with retries, so things need to be adjusted, or we should use a decorator here too, it is more universal anyway.

File: pydrive2/test/test_util.py [hunk 24]
```diff
@@ -1,7 +1,55 @@
...
+
+def pydrive_list_item(drive, query, max_results=1000):
+    param = {"q": query, "maxResults": max_results}
+
+    file_list = drive.ListFile(param)
+
+    # Isolate and decorate fetching of remote drive items in pages
+    # Fetch pages until None is received, lazily flatten the thing
+    return cat(
+        iter(lambda: pydrive_retry(lambda: next(file_list, None)), None)
```
[the remark this answers is not in the document]
Comment (Suor) [reply pydrive2/test/test_util.py]: Also, combining this into one line made it more hard to comprehend. It used to be comment - code - comment - code, less to bite at a time.

A separate thought - maybe it is a good idea to expose this for users not only use in tests.

**Comment by MaxRis** [comment]
going to backup code formatting changes in remote branch `enable-code-formatting-checks`, will remove these changes from this PR

File: pydrive2/test/test_apiattr.py (hunk 18 above)
Comment (MaxRis) [reply pydrive2/test/test_apiattr.py]: thanks, works good without lambda as well; Will simplify for all places without params in function call.

File: pydrive2/test/test_apiattr.py (hunk 19 above)
Comment (MaxRis) [reply pydrive2/test/test_apiattr.py]: > why don't we use GoogleDriveTest here?

original auth implementation is not common for all automated TestCase classes. For example, it might be:

```
@classmethod
    def setup_class(cls):
        setup_credentials()
```

or

```
@classmethod
    def setup_class(cls):
        setup_credentials()

        cls.ga = GoogleAuth('pydrive2/test/settings/default.yaml')
        cls.ga.ServiceAuth()
```

I've stopped refactoring there on common logic moving to the `setup_credentials()` defined in utils.

File: pydrive2/test/test_oauth.py (hunk 20 above)
Comment (MaxRis) [reply pydrive2/test/test_oauth.py]: thank you for finding this, this is fixed by renaming `test1.yaml` to `test_oauth_test_01` and reverting to the old test1.yaml content.

File: pytest.ini (hunk 21 above)
Comment (MaxRis) [reply pytest.ini]: good idea, I've added marker `manual` and removed `automated`.

File: pydrive2/test/test_util.py (hunk 22 above)
Comment (MaxRis) [reply pydrive2/test/test_util.py]: it seems I had an assumption that Travis runs parallel jobs on the same filesystem which can cause such errors pretty often, but, thanks, I will remove `except` to let it error out.

File: pydrive2/test/test_util.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (MaxRis) [reply pydrive2/test/test_util.py]: @Suor 

> It's a bit confusing that gdrive_retry() is a decorator in dvc and pydrive_retry() does an immediate action here. Also, we only use it in tests here I suppose? 

Yes, `pydrive_retry` is used only for tests at the moment. API request errors started to appear only on Travis ( 10 parallel jobs for PR reach user API usage quota very easy ), but locally it is not possible to observe the errors. 

> Letting a user handle retries, this is a design decision?
A separate thought - maybe it is a good idea to expose this for users not only use in tests.

Looks like it is original design decision. But pretty strange to see that retry logic was added at lower level to `google-api-python-client` package.
Would be great to provide `retry` functionality to the users of PyDrive, I think.

> Also, combining this into one line made it more hard to comprehend. It used to be comment - code - comment - code, less to bite at a time.

There was warning to not assign lambda to variable, but at this time I was not able to find better workaround. For now, I've reverted commit with formatting checks, so the implementation is 2 lines old version copy/pasted from DVC.

**Comment by MaxRis** [comment]
@shcheklein @Suor could you please check if it looks good now, thanks

File: pydrive2/test/test_apiattr.py (hunk 19 above)
Comment (shcheklein) [reply pydrive2/test/test_apiattr.py]: 👍

File: setup.py [hunk 25]
```diff
@@ -20,6 +20,6 @@
         "pyOpenSSL >= 19.1.0"
     ],
     extras_require={
-        "tests": ["timeout-decorator"],
+        "tests": ["pytest>=4.6.0,<5.1.2", "timeout-decorator", "funcy>=1.14,<2"],
```
Comment (shcheklein) [inline setup.py]: @Suor @MaxRis let's not do < version. It's better to fail fast and fix the discrepancy, then being not aware about new versions. We'll always have time to do this.
